common/src/main/java/com/wynntils/wynn/model/CompassModel.java

```diff
@@ -5,39 +5,73 @@
 package com.wynntils.wynn.model;
 
 import com.wynntils.core.managers.Model;
+import com.wynntils.mc.event.ClientTickEvent;
 import com.wynntils.mc.event.SetSpawnEvent;
 import com.wynntils.mc.objects.Location;
 import com.wynntils.mc.utils.McUtils;
 import com.wynntils.wynn.model.map.poi.MapLocation;
 import com.wynntils.wynn.model.map.poi.WaypointPoi;
 import java.util.Optional;
+import java.util.function.Supplier;
 import net.minecraft.core.BlockPos;
 import net.minecraftforge.eventbus.api.SubscribeEvent;
 
 public final class CompassModel extends Model {
+    private static Supplier<Location> supplier = null;
     private static Location compassLocation = null;
 
     public static void init() {}
 
+    @SubscribeEvent
+    public void onTick(ClientTickEvent.Start e) {
+        if (supplier == null) return;
+
+        Location newLocation = supplier.get();
+
+        if (newLocation == null) { // drop location
+            reset();
+        } else if (compassLocation != newLocation) { // update location
+            compassLocation = newLocation;
+
+            if (McUtils.mc().level != null) {
+                McUtils.mc().level.setDefaultSpawnPos(compassLocation.toBlockPos(), 0);
+            }
+        }
+    }
+
     public static Optional<Location> getCompassLocation() {
         return Optional.ofNullable(compassLocation);
     }
 
     public static Optional<WaypointPoi> getCompassWaypoint() {
         if (compassLocation != null) {
-            Location location = CompassModel.getCompassLocation().get();
             // Always render waypoint POI on top
-            WaypointPoi waypointPoi =
-                    new WaypointPoi(new MapLocation((int) location.x, Integer.MAX_VALUE, (int) location.z));
+            WaypointPoi waypointPoi = new WaypointPoi(
+                    new MapLocation((int) compassLocation.x, Integer.MAX_VALUE, (int) compassLocation.z));
 
             return Optional.of(waypointPoi);
         }
 
         return Optional.empty();
     }
 
-    public static void setCompassLocation(Location compassLocation) {
-        CompassModel.compassLocation = compassLocation;
+    public static void setDynamicCompassLocation(Supplier<Location> compassSupplier) {
```

Comment:
Why don't we use our new Poi interface and implement WaypointPoi as a "special" POI? That way you can get rid of this supplier weirdness and directly support it in the class.